_posts/2016-05-22-th1eadin0g1.markdown

```diff
@@ -0,0 +1,226 @@
+title: "Threading 101 for iOS"
+date: 2016-05-22 12:09
+author: orta
+categories: [mobile, cocoapods, newbies]
+---
+
+Once an app has reached a level of complexity wherein Apple would _not_ [reject it](https://developer.apple.com/app-store/review/rejections/) for "Not enough lasting value" it has probably had to deal with multiple thread execution.
+
+This could be as simple as downloading a file, or image from the internet, accessing an external API, using something like background locations or playing sounds. There are no limits to the amount of cases where you could be doing some asynchronous work under the hood at the same time. So, I'm going to try and cover _most_ of the common use cases of where you are using it without knowing, then expand on a few techniques to write useful multi-threaded code.
+
+<!-- more -->
+
+### Thread Terminology 101
+
+* `Process` - The context within an app is running, you will have one process per app. 
+* `Thread` - A pathway of code, you can have multiple threads in an app's process.
+* `Main Thread` - The initial thread for your app, this is the only thread where you can do user interface code.
+* `Blocked` - You're waiting for some work to happen on a thread, so that something else can happen.
+* `Deadlock` - One thread is waiting for another thread to finish, but that thread is waiting for the other. Potentially either finish waiting.
+* `Mutable State` - State is the name for keeping track of the somethinge: e.g. a variable. Mutable state is keeping track of something that can also change. 
+* `Immutable state` - Immutability is keeping track of something, and knowing it won't change.
+
+### "Same Time"
+
+Saying that something is happening at the same time, is potentially a lie, or more accurately a simplification. I'm currently writing this post on a MacBook Pro, with these specs:
+
+```
+Hardware Overview:
+
+  Model Name:	MacBook Pro
+  Model Identifier:	MacBookPro12,1
+  Processor Name:	Intel Core i7
+  Processor Speed:	3.1 GHz
+  Number of Processors:	1
+  Total Number of Cores:	2
+  L2 Cache (per Core):	256 KB
+  L3 Cache:	4 MB
+  Memory:	16 GB
+```
+
+Which means there's one processor with two cores. This means my processor has [two processing](https://en.wikipedia.org/wiki/Multi-core_processor) units. However, a processor can also fake being more than one processor at once. This is known as [Hyperthreading](https://en.wikipedia.org/wiki/Hyper-threading) which is a tool used to help improve the faking of multiple processors. Effectively, it can create virtual processor cores. This means I can run two sets of processing at once on this computer.
+
+If I was using a single-core processor, it wouldn't be possible to do two simultaneous instruction. However, because a processor can do work so fast, it can jump between the sets of instructions, that it's imperceptual to a human. So even if there's only one processor, it's possible to do multi-threading to give the impression of background work.
```

Comment:
It's unclear how multi processor execution maps to threads. Add something to the effect of a thread being that series of instructions executing on a specific processors. Then when another thread is executing on the same processor it's called a context switch.